### PR TITLE
HPCC-21289 Refactor the disk io and network stats

### PR DESCRIPTION
**Pull request by ghalliday**
Signed-off-by: Gavin Halliday <gavin.halliday@lexisnexis.com>

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [x] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

### CODE REVIEW

**Comment by ghalliday** [comment]
@mckellyln please review.  Apologies for another refactoring PR!

**Comment by richardkchapman** [comment]
@ghalliday Smoketest is failing

**Comment by AttilaVamos** [comment]
@AttilaVamos Those spray errors has been caused by my merged PR-https://github.com/hpcc-systems/HPCC-Platform/pull/12206 . I'm working on the fix. I am sorry about inconvenience caused.

File: system/jlib/jdebug.cpp [hunk 1]
```diff
@@ -1706,61 +1754,360 @@ class CProcessMonitor
...
+    return true;
+}
+MODULE_EXIT()
+{
+    delete globalOsDiskInfo.queryExisting();
+}
+
+const OsDiskInfo & queryGlobalOsDiskInfo()
+{
+    return *globalOsDiskInfo.query([] { return new OsDiskInfo; });
```
Comment (mckellyln) [inline system/jlib/jdebug.cpp]: I need to understand this line

**Review by mckellyln — COMMENTED** [review]
Overall looks very good.
I just want to fully understand these statements:
```
static Singleton<OsDiskInfo> globalOsDiskInfo;
MODULE_EXIT()
{
    delete globalOsDiskInfo.queryExisting();
}

const OsDiskInfo & queryGlobalOsDiskInfo()
{
    return *globalOsDiskInfo.query([] { return new OsDiskInfo; });
}
```
@ghalliday

File: system/jlib/jdebug.cpp (hunk 1 above)
Comment (ghalliday) [reply system/jlib/jdebug.cpp]: It is a way of implementing code like the following:

  if (!ptr)
  {
     CriticialBlock block(cs);
     if (!ptr)
        ptr = new OsDiskInfo;
  }
  return ptr;

ensuring that atomics are used correctly and not re-evaluated.

**Comment by ghalliday** [comment]
The module exit code is there to clean up on exit - if it has been created.

It was implemented to create on first use to avoid overhead when not needed and unnecessary tracing.

**Comment by richardkchapman** [comment]
I think it's the weird singleton syntax Mark was querying. Using a c++11 lambda I think

**Comment by ghalliday** [comment]
yes.  @mckellyln IM me to discuss when you get back

**Comment by mckellyln** [comment]
ok, after reading some more about lambda functions I understand it.

**Review by mckellyln — APPROVED** [review]

**Pull request merged**